capstone/src/main/java/com/google/sps/servlets/TagsTFIDFServlet.java

```diff
@@ -144,21 +192,59 @@ public void calculateTFIDF(LinkedHashMap<Long, LinkedHashMap<String, Integer>> g
     }
   }
 
+  /**
+   * Weights the tf-idf score of a given ngram by taking into account its length.
+   */
+  private double weightScore(double score, String ngram) {
+    double wordsWeight = (ngram.split(" ").length / 3.1);
+    double lengthWeight = ((double) ngram.length() / longestNgramLength);
+
+    double totalWeight = 1 + wordsWeight + lengthWeight;
+    score *= totalWeight;
+
+    return score;
+  }
+
   /**
    * Given a priority queue of tags, stores the top 3 tag values in the Group's datastore.
    */
   public void putTagsInDatastore(PriorityQueue<Tag> tagQueue, long groupId, 
       HttpServletResponse response) throws IOException {
-
+    
     ArrayList<Tag> topTags = new ArrayList<>();
     int tagsLength = (tagQueue.size() >= 3) ? 3 : tagQueue.size();
     for (int i = 0; i < tagsLength; i++) {
-      topTags.add(tagQueue.poll());
+      Tag next = tagQueue.poll();
+      if (next != null) {
+        if (checkIfDuplicate(topTags, next)) {
+          i--;
+        } else {
+          topTags.add(next);
+        }
+      }
     }
     ArrayList<EmbeddedEntity> tags = Tag.createTagEntities(topTags);
 
     Entity groupEntity = ServletHelper.getEntityFromId(response, groupId, datastore, "Group");
     groupEntity.setProperty("tags", tags);
     datastore.put(groupEntity);
   }
+
+  /**
+   * Checks if a potential group tag is redundant.
+   */
+  public boolean checkIfDuplicate(ArrayList<Tag> topTags, Tag next) {
+    for (Tag tag : topTags) {
+      String[] existingWords = tag.getText().split(" ");
+      String[] nextWords = next.getText().split(" ");
+      for (String existingWord : existingWords) {
+        for (String word : nextWords) {
+          if (word.equals(existingWord)) {
+            return true;
+          }
+        }
+      }
+    }
```

Comment:
instead of O(n**3), consider making a Set<String> of existing words, so you don't have to iterate over all existing terms and loop over them all. This could instead just be 

```java
for (String term in next.split(" ")) {
  if (existingTerms.contains(term)) {
    return true;
  }
}
return false;

Reply:
This would work as an optimization if I didn't change any of my deduplication logic. If I did change the logic like how I showed in the comment above, then I'm wouldn't be able to use the set.

Reply:
Ack